### PR TITLE
Warn if positional arguments are used in MetricFrame

### PR DESCRIPTION
**Pull request by MiroDudik**
The initial step in #756. Specifically, Step 1 from https://github.com/fairlearn/fairlearn/issues/756#issuecomment-843235220, but with the benefit of the constructor signature that looks like Step 2.

It's currently using `_deprecate_positional_args` utility from sklearn ([see here](https://github.com/scikit-learn/scikit-learn/blob/15a949460dbf19e5e196b8ef48f9712b72a3b3c3/sklearn/utils/validation.py#L35)).

This is what the warning looks like:
```pycon
>>> from fairlearn.metrics import MetricFrame
>>> import sklearn.metrics as skm
>>> y_true=[0,1,0,0,1]
>>> y_pred=[0,1,1,0,1]
>>> sf=['a','a','b','b','b']
>>> mf = MetricFrame(skm.accuracy_score, y_true, y_pred, sensitive_features=sf)
/mnt/c/Users/mdudik/projects/fairlearn/.venv/lib/python3.6/site-packages/sklearn/utils/validation.py:72: FutureWarning: Pass metrics=<function accuracy_score at 0x7fe2a3e45598>, y_true=[0, 1, 0, 0, 1], y_pred=[0, 1, 1, 0, 1] as keyword args. From version 0.8.0 passing these as positional arguments will result in an error
  "will result in an error", FutureWarning)
>>> mf.difference()
0.33333333333333337
>>> mf = MetricFrame(metrics=skm.accuracy_score, y_true=y_true, y_pred=y_pred, sensitive_features=sf)
>>> mf.difference()
0.33333333333333337
```

I don't love the look of the warning message, but I like the convenience of just using the decorator to add the warning to the code.

Signed-off-by: MiroDudik <mdudik@gmail.com>

### CODE REVIEW

**Review by riedgar-ms — APPROVED** [review]
No point recreating a warning that already exists. If it's not out by v0.8.0, we should remember to revise the message, though ;-)

**Review by romanlutz — COMMENTED** [review]
Please add a note to CHANGES.md for this.

**Comment by MiroDudik** [comment]
@hildeweerts : thanks for fixing the example! After this PR is in, I was planning to create a separate issue to fix all the examples / notebooks / tests. Let me know if any other updates are required at this point besides updating CHANGES.md as roman requested.

**Comment by hildeweerts** [comment]
Would it make sense to add a more meaningful error message for users who use `metric` instead of `metrics`, or is that not necessary in your opinion?

**Comment by romanlutz** [comment]
Theoretically one could keep it as a kwarg for a release or two and map it back to the new metrics, while printing a warning. I'm not sure that's needed, though. I'd be curious to hear what @riedgar-ms amd @adrinjalali think, too.

**Comment by hildeweerts** [comment]
I was thinking about catching the `TypeError: MetricFrame got an unexpected keyword argument metric` (or whatever the basic error looks like) to say something more meaningful like "did you mean `metrics`?" or "`metric was changed to metrics in version blabla`.

I agree it's probably not necessary to actually map it back, but I'm curious about other people's thoughts!

**Comment by MiroDudik** [comment]
@hildeweerts , @romanlutz : I don't want to mess with the function signature, but I could see whether I can create a custom decorator that would basically have an additional keyword argument "metric" and throw a less obscure warning if that argument is used.

**Review by romanlutz — APPROVED** [review]
LGTM assuming you add a note to CHANGES.md as you mentioned.

**Comment by MiroDudik** [comment]
I've added a custom decorator. The behavior is now this:

```pycon
>>> from fairlearn.metrics import MetricFrame
>>> import sklearn.metrics as skm
>>> y_true=[0,1,0,0,1]
>>> y_pred=[0,1,1,0,1]
>>> sf=['a','a','b','b','b']

>>> mf = MetricFrame(skm.accuracy_score, y_true, y_pred, sensitive_features=sf)
/mnt/c/Users/mdudik/projects/fairlearn/fairlearn/metrics/_metric_frame.py:67: FutureWarning: Pass metrics=<function accuracy_score at 0x7f2ce8c87ea0>, y_true=[0, 1, 0, 0, 1], y_pred=[0, 1, 1, 0, 1] as keyword args. From version 0.8.0 passing these as positional arguments will result in an error
  FutureWarning)
>>> mf.difference()
0.33333333333333337

>>> mf = MetricFrame(skm.accuracy_score, y_true=y_true, y_pred=y_pred, sensitive_features=sf)
/mnt/c/Users/mdudik/projects/fairlearn/fairlearn/metrics/_metric_frame.py:67: FutureWarning: Pass metrics=<function accuracy_score at 0x7f2ce8c87ea0> as keyword args. From version 0.8.0 passing these as positional arguments will result in an error
  FutureWarning)
>>> mf.difference()
0.33333333333333337

>>> mf = MetricFrame(metric=skm.accuracy_score, y_true=y_true, y_pred=y_pred, sensitive_features=sf)
/mnt/c/Users/mdudik/projects/fairlearn/fairlearn/metrics/_metric_frame.py:77: FutureWarning: The positional argument 'metric' has been replaced by a keyword argument 'metrics'. From version 0.8.0 passing it as a positional argument or as a keyword argument 'metric' will result in an error
  FutureWarning)
>>> mf.difference()
0.33333333333333337

>>> mf = MetricFrame(metrics=skm.accuracy_score, y_true=y_true, y_pred=y_pred, sensitive_features=sf)
>>> mf.difference()
0.33333333333333337
```

@hildeweerts , @romanlutz , @riedgar-ms : please review that you're still ok with this!

**Review by hildeweerts — APPROVED** [review]

**Review by riedgar-ms — CHANGES_REQUESTED** [review]
Although I'm happy with the changes so far, I've just remembered that the docs and examples need updating too (not to mention the tests). We shouldn't have warnings being emitted from those.

**Comment by MiroDudik** [comment]
@riedgar-ms -- i am aware and agree, see my comment https://github.com/fairlearn/fairlearn/pull/822#issuecomment-848766235. let me fix the examples and docs and leave tests and notebooks for sprints.

**Review by romanlutz — COMMENTED** [review]
> let me fix the examples and docs and leave tests and notebooks for sprints.

I'm not quite sure that's how we should be treating these. I know you wrote this in the context of creating good first issues (in fact I asked everyone to think about this a bit yesterday!). However, we set criteria for PRs such as having documentation and tests. If we start now offloading supposedly simple test adjustments to the future then how can we ask other contributors to write tests?

We already have a task to fix warnings in tests. Unsurprisingly nobody has picked that up because it's just not terribly interesting. Besides, sometimes these can take a while to investigate. I'd vote for not knowingly adding more, but I'm sure other @fairlearn/fairlearn-maintainers have opinions, too.

**Comment by MiroDudik** [comment]
Well... I did one pass with a couple of RegEx queries. `.ipynb` notebooks are incomplete, but I'm hoping it covered the rest. And as a bonus, I think I fixed a FutureWarning in exponentiated gradient tests.

**Review by romanlutz — APPROVED** [review]
Thanks for the test update. I don't care as much about the ipynb notebooks, more about examples, so this is fine by me.

File: CHANGES.md
Comment (riedgar-ms) [inline CHANGES.md:7]: Could this description feature 'keyword arguments' somewhere?

File: docs/user_guide/mitigation.rst
Comment (riedgar-ms) [inline docs/user_guide/mitigation.rst:615]: Small thing, but if an argument list is being split over multiple lines, I usually like to keep one argument per line (opinions on this may differ, and I'm not wholly consistent either)

**Review by riedgar-ms — APPROVED** [review]
Thank you for eliminating the warnings from our own code.

**Review by adrinjalali — CHANGES_REQUESTED** [review]
Thanks @MiroDudik

File: CHANGES.md
Comment (MiroDudik) [reply CHANGES.md:7]: will do!

File: docs/user_guide/mitigation.rst
Comment (MiroDudik) [reply docs/user_guide/mitigation.rst:615]: i can do this... this reminds me... do we want to switch to black over flake8? pros/cons?

File: fairlearn/metrics/_metric_frame.py
Comment (MiroDudik) [reply fairlearn/metrics/_metric_frame.py:63]: @adrinjalali : this is actually almost exactly what [sklearn does](https://github.com/fairlearn/fairlearn/pull/822#issue-652386593). i don't like it that much either :-)... how about this:
```pycon
>>>
>>> mf = MetricFrame(skm.accuracy_score, y_true, y_pred, sensitive_features=sf)
FutureWarning: You have provided metrics, y_true, y_pred as positional arguments. Please pass them as keyword arguments. From version 0.8.0 passing these as positional arguments will result in an error.
```

File: fairlearn/metrics/_metric_frame.py
Comment (MiroDudik) [reply fairlearn/metrics/_metric_frame.py:177]: Not sure I quite understand. The code in this PR issues a deprecation warning if you pass `metric` (it's handled by the decorator). You get the following message:
```pycon
>>> mf = MetricFrame(metric=skm.accuracy_score, y_true=y_true, y_pred=y_pred, sensitive_features=sf)
FutureWarning: The positional argument 'metric' has been replaced by a keyword argument 'metrics'. From version 0.8.0 passing it as a positional argument or as a keyword argument 'metric' will result in an error
```

File: docs/user_guide/mitigation.rst
Comment (adrinjalali) [reply docs/user_guide/mitigation.rst:615]: to me, black does the formatting, flake8 tests the compliance.

File: fairlearn/metrics/_metric_frame.py
Comment (adrinjalali) [reply fairlearn/metrics/_metric_frame.py:177]: I see. I would rather not have this warning in the decorator. The decorator is nice to handle the positional argument deprecation, but we can easily deprecate an old parameter in the function itself, which would make the code more readable.

File: fairlearn/metrics/_metric_frame.py
Comment (adrinjalali) [reply fairlearn/metrics/_metric_frame.py:63]: I like your proposed alternative.

File: fairlearn/metrics/_metric_frame.py
Comment (MiroDudik) [reply fairlearn/metrics/_metric_frame.py:177]: I see. If I understand correctly, you'd like to have both `metrics` and `metric` as keyword arguments, but:
* in the numpy doc section use a [deprecation block](https://numpydoc.readthedocs.io/en/latest/format.html#deprecation-warning)
* in the constructor code, issue a warning

Comment (MiroDudik) [reply fairlearn/metrics/_metric_frame.py:177]: If both `metrics` and `metric` are passed, raise `TypeError` or issue a warning and ignore `metric`?

Comment (adrinjalali) [reply fairlearn/metrics/_metric_frame.py:177]: Yes, exactly

**Comment by adrinjalali** [comment]
If both `metric` and `metrics` are passed, I'm happy to raise.

File: fairlearn/metrics/_metric_frame.py
Comment (MiroDudik) [reply fairlearn/metrics/_metric_frame.py:177]: Actually, there's an issue--if we want to have compatibility with the old API, we need to make `metrics` / `y_true` / `y_pred` formally optional, otherwise we'll run into this:
```pycon
>>> mf = MetricFrame(metric=skm.accuracy_score, y_true=y_true, y_pred=y_pred, sensitive_features=sf)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mnt/c/Users/mdudik/projects/fairlearn/fairlearn/metrics/_metric_frame.py", line 71, in compatible_metric_frame_init
    **kwargs)
TypeError: __init__() missing 1 required keyword-only argument: 'metrics'
```
So my suggestion is to have the signature where `metrics` / `y_true` / `y_pred` are optional, but raise if any of them is `None`. Thoughts?

Comment (MiroDudik) [reply fairlearn/metrics/_metric_frame.py:190]: @adrinjalali : when I tried to do this without a decorator, I found myself re-implementing standard Python error warnings around missing keyword-only arguments, so I've decided to go with the decorator approach, but to add `metric=None` in the signature and docs. (maybe that makes code actually more confusing?)

**Comment by MiroDudik** [comment]
@adrinjalali -- i'm done with my updates. i ended up keeping `metric` keyword handling in the decorator, but added a comment and a deprecation block to address confusion.

**Review by adrinjalali — CHANGES_REQUESTED** [review]
I guess since this is gonna be all deleted soon, I'm happy with the implementation. However, since we're doing releases rather quickly, I'd let it be there for at least one extra release.

Could you please write tests for the changes? To make sure the right warnings are raised when planned and no warning is raised when correctly used?

We need codecov, gonna open a separate issue for that.

**Comment by MiroDudik** [comment]
@adrinjalali -- I've added tests and also changed the deprecation warning, so it says that the old API will raise errors starting in 0.9.0. (The next version we're about to release is 0.7.0.) Is that what you had in mind?

**Review by adrinjalali — APPROVED** [review]
LGTM. I'd go for deprecation end around v.10 instead of .9 though. We do tend to release rather frequently.

**Comment by MiroDudik** [comment]
i've bumped the future version from 0.9.0 to 0.10.0. once the checks pass, please merge! @riedgar-ms @adrinjalali

**Pull request merged**
